libs/s25main/mapGenerator/HeadQuarters.h

```diff
@@ -42,45 +43,71 @@ namespace rttr { namespace mapGenerator {
      * @param area area within the HQ position should be
+     * @param playerDistanceToMountains desired player distance to mountain
      *
      * @return all suitable HQ positions witihn the specified area (or the entire map if the area is empty).
      * @throw runtime_error
      */
     template<class T_Container>
-    std::vector<MapPoint> FindHqPositions(const Map& map, const T_Container& area)
+    std::vector<MapPoint> FindHqPositions(const Map& map, const T_Container& area, unsigned playerDistanceToMountains)
     {
         if(area.empty())
         {
             throw std::runtime_error("Could not find any valid HQ position!");
         }
 
-        const auto isValid = [](const MapPoint& pt) { return pt.isValid(); };
-        const auto isObstacle = [&map](MapPoint point) {
-            return map.textureMap.Any(point, [](auto t) { return !t.Is(ETerrain::Buildable); });
+        const auto isObstacle = [&map](const MapPoint& pt) {
+            return map.textureMap.Any(pt, [](auto t) { return !t.Is(ETerrain::Buildable); });
         };
-        std::vector<MapPoint> hqs;
-        std::copy_if(map.hqPositions.begin(), map.hqPositions.end(), std::back_inserter(hqs), isValid);
 
-        // Quality of a map point as one player's HQ is a mix of:
-        // 1. distance to other players' HQs (higher = better)
-        // 2. distance to any other obstacle e.g. mountain & water (higher = better)
-        NodeMapBase<unsigned> potentialHqQuality = DistancesTo(hqs, map.size);
+        // Ensure possible HQ positions are far enough from any obstacles
         const auto obstacleDistance = DistancesTo(map.size, isObstacle);
-        const auto minDistance = helpers::clamp(GetMaximum(obstacleDistance, area), 2u, 4u);
+        const auto minObstacleDistance = helpers::clamp(GetMaximum(obstacleDistance, area), 2u, 4u);
+        const auto farFromObstacles = [&obstacleDistance, minObstacleDistance](const MapPoint& pt) {
+            return obstacleDistance[pt] >= minObstacleDistance;
+        };
+        std::vector<MapPoint> possiblePositions;
+        std::copy_if(area.begin(), area.end(), std::back_inserter(possiblePositions), farFromObstacles);
+
+        if(possiblePositions.empty())
+        {
+            return possiblePositions;
+        }
+
+        // Ensure HQ positions keep desired distance to mountains
+        const auto mountain = [&map](const MapPoint& pt) { return map.textureMap.Any(pt, IsMinableMountain); };
+        const auto mountainDistance = DistancesTo(map.size, mountain);
+        const auto desiredDistance =
+          std::max(GetMinimum(mountainDistance, possiblePositions), playerDistanceToMountains);
+        const auto maxDistance = static_cast<unsigned>(map.size.x + map.size.y) / 4;
 
         std::vector<MapPoint> positions;
-        for(const MapPoint& pt : area)
+        auto allowedMountainDistance = desiredDistance;
+        while(positions.empty() && allowedMountainDistance < maxDistance)
```

Comment:
I fear that this becomes quite expensive. For a 512x512 map this may run 256 times over possiblePositions

An idea I have: Sort possiblePositions by mountainDistance. Then you can determine the range in that sequence and move that, this way you basically go over possiblePositions (after the sort) at most once.

However thinking about this: Why does the range only increase to upper bound and moves away from lower? Is that intended? E.g. for `allowedMountainDistance=10` it will check points with distance in '(5, 15)` (exclusive), then '(6 , 16)`, ... If this is intended, then the above approach becomes even easier: If the first range did not find anything, you simply need to check if there are points left and if so take the next one and all with equal distance.

Reply:
Absolutely right. I changed the whole thing a bit now to widen the search radius rather than changing the `allowedMountainDistance`. Also I didn't consider distance between multiple HQs at all, so I added a minimum distance as well.